### PR TITLE
Btree refactor

### PR DESCRIPTION
**Pull request by boki1**
# Description
Refactor current implementation in order to improve in two major areas:
    1. Limited amount of elements in the tree.
    2. Improved performance

# Test suite
Insert a lot of elements (as of now that is 200_000, but the test has been done with up to 500_000).

# Environment
Using gcc (GCC) 11.1.0 on a x86_64 machine with GNU/Linux.

--------------------

#### This PR makes changes which ...

- [x] are thoroughly tested
- [ ] are well documented

### CODE REVIEW

File: src/core/storage/Page.cpp [hunk 1]
```diff
@@ -21,7 +25,7 @@ TEST_CASE("Page", "[page]") {
 	std::array<uint8_t, Page::size()> p3_data;
 	std::iota(p3_data.begin(), p3_data.end(), 3);
 	Page p3(std::move(p3_data));
```
Comment (StoyanTinchev) [inline src/core/storage/Page.cpp:27]: Are you sure that you need to std::move the "data" here and then again in the constructor

Comment (StoyanTinchev) [reply src/core/storage/Page.cpp:27]: Also you do this from "p" to "p3"

File: src/core/storage/Page.cpp [hunk 2]
```diff
@@ -2,12 +2,16 @@
...
+	std::ofstream of{std::string{fname}, std::ios::trunc};
+}
+
 using namespace internal::storage;
 
 TEST_CASE("Page", "[page]") {
+	truncate_file("/tmp/eu-pager");
 	Pager pr("/tmp/eu-pager");
 
 	std::array<uint8_t, Page::size()> p1_data;
```
Comment (StoyanTinchev) [inline src/core/storage/Page.cpp:17]: This and the others are not initialised - probably seg fault?

File: src/core/storage/Page.h [hunk 3]
```diff
@@ -17,11 +20,15 @@ class Pager;
...
 class Page {
-	static inline constexpr int32_t SIZE = 1 << 14;
+	static inline constexpr int32_t SIZE = 4_KB;
 
 	friend Pager;
 
 public:
+	Page() : m_data{0} {}
+
+	static Page empty() noexcept { return Page(); }
```
Comment (StoyanTinchev) [inline src/core/storage/Page.h:30]: `return {};` :)

File: src/core/storage/PageCache.h [hunk 4]
```diff
@@ -64,41 +64,50 @@ class PageCache {
...
-			evict_lru();
+			evict();
 	}
 
 	[[nodiscard]] Position get_new_pos() noexcept { return m_pager.alloc(); }
 
-	[[nodiscard]] bool full() const noexcept { return m_index.size() == m_limit; }
+	[[nodiscard]] bool full() const noexcept { return m_index.size() >= m_limit; }
 
 	[[nodiscard]] bool empty() const noexcept { return m_index.size() == 0; }
```
Comment (StoyanTinchev) [inline src/core/storage/PageCache.h]: `m_index.empty()`

File: src/core/storage/Position.h [hunk 5]
```diff
@@ -1,12 +1,19 @@
...
 #include <string>
 
 namespace internal::storage {
 
 class Position {
 	friend std::hash<Position>;
+
 public:
 	Position() = default;
 	Position(const Position &) = default;
```
Comment (StoyanTinchev) [inline src/core/storage/Position.h:19]: `explicit Position(long pos) : m_pos{pos}, m_isset{true} {}`
`explicit operator long() const { return m_pos; }`
in 26 and 27 rows

File: src/core/storage/btree/Bench.cpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
+#include <random>
+
+#include <benchmark/benchmark.h>
+
+#include <core/storage/Storage.h>
```
Comment (StoyanTinchev) [inline src/core/storage/btree/Bench.cpp]: Unused

File: src/core/storage/btree/Btree.cpp [hunk 7]
```diff
@@ -1,67 +1,250 @@
...
+
+template<typename T>
+T item();
+
+template<>
+int item<int>() {
+	static std::random_device dev;
+	static std::mt19937 rng(dev());
+	static std::uniform_int_distribution<std::mt19937::result_type> dist(1, 10000000);
+	return dist(rng);
```
Comment (StoyanTinchev) [inline src/core/storage/btree/Btree.cpp:26]: Conversion from `unsigned int` to `signed int` -> possible segmentation fault

File: src/core/storage/btree/Node.cpp [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+	auto bn = Nod(Metadata(Branch({}, {})), {}, false); auto &b = bn.branch();
+	auto ln = Nod(Metadata(Leaf({}, {})), 13, true); auto &l = ln.leaf();
+
+	constexpr auto limit_branch = 512;
+	constexpr auto limit_leaf = 512;
+
+	// Fill
+	for (uint32_t i = 0; i < limit_branch; ++i) {
+		b.m_refs.push_back(i);
+		b.m_links.push_back(Position(i));
```
Comment (StoyanTinchev) [inline src/core/storage/btree/Node.cpp]: `emplace_back`? and at 145 row too

**Review by StoyanTinchev — CHANGES_REQUESTED** [review]
I noticed a few places where you use `unsigned` type but `signed` is required. Probably they are your hidden segmentation fault problem ;)

File: src/core/storage/Page.cpp (hunk 2 above)
Comment (boki1) [reply src/core/storage/Page.cpp:17]: What do you mean not initialized. The default constructor is called.

File: src/core/storage/Page.cpp (hunk 1 above)
Comment (boki1) [reply src/core/storage/Page.cpp:27]: I am not sure that I understand you correctly. Could you clarify please?

What I am intending is to create a `std::array` of size `Page::size()` and fill it with the value 3. Then, these values are moved inside `p3` when calling its constructor which accepts `std::array<> &&`.

File: src/core/storage/Page.h (hunk 3 above)
Comment (boki1) [reply src/core/storage/Page.h:30]: Yes, that would do the job as well.

File: src/core/storage/Position.h (hunk 5 above)
Comment (boki1) [reply src/core/storage/Position.h:19]: No. I am intentionally making this implictly convertible to `long`.

File: src/core/storage/btree/Btree.cpp (hunk 7 above)
Comment (boki1) [reply src/core/storage/btree/Btree.cpp:26]: No. Segmentation faults cannot be caused to using a type with different sign.

File: src/core/storage/btree/Node.cpp (hunk 8 above)
Comment (boki1) [reply src/core/storage/btree/Node.cpp]: Yes. That would be better.

**Pull request merged**